### PR TITLE
Extended ID permissions supported by bidder

### PR DESCRIPTION
**Pull request by YerkovichM**
## Type of change

- [x] Feature

## Description of change

#5814 
Bid adapters receive only allowed user ids

**Case 1**: 
```
userSync: {
 userIds: [
  {
    name: "sharedId",
    bidders: [
      'rubicon',
      'exBidder'
    ],
    params: {
      syncTime: 60 //inseconds,
      defaultis24hours
    },
    storage: {
      type: "cookie",
      name: "sharedid",
      expires: 28
    }
  }
]
}
```

Rubicon bid adapter receives:
```
{
  "bidder": "rubicon",
  ...
  "userId": {
    "sharedid": {
      "id": "01EQDMBY1YPFFVY57R2V11PY89",
      "third": "01EQDMBY1YPFFVY57R2V11PY89"
    }
  },
  "userIdAsEids": [
    {
      "source": "sharedid.org",
      "uids": [
        {
          "id": "01EQDMBY1YPFFVY57R2V11PY89",
          "atype": 1,
          "ext": {
            "third": "01EQDMBY1YPFFVY57R2V11PY89"
          }
        }
      ]
    }
  ],
  ...
}
```
PrebidServerBidAdapter request sent as:
```
{
  ...
  "ext": {
    "prebid": {
      ...
      "data": {
        "eidpermissions": [
          {
            "source": "sharedid.org",
            "bidders": [
              "rubicon"
            ]
          }
        ]
      }
    }
  },
  ...
  "user": {
    "ext": {
      "eids": [
        {
          "source": "sharedid.org",
          "uids": [
            {
              "id": "01EQDMBY1YPFFVY57R2V11PY89",
              "atype": 1,
              "ext": {
                "third": "01EQDMBY1YPFFVY57R2V11PY89"
              }
            }
          ]
        }
      ]
    }
  }
}
```
**Case 2**:
```
userSync: {
  userIds: [{
    name: "sharedId",
    bidders: ['exBidder'],
    ...
    }
  ]
}
```
Rubicon bid adapter receives:
```
{
  "bidder": "rubicon",
  ... no user ID entries ...
}
```
PrebidServerBidAdapter request:
```
{
  ...
  "ext": {
    "prebid": {
      ...
      "data": {
        "eidpermissions": [
          {
            "source": "sharedid.org",
            "bidders": [ 'exBidder' ]
          }
        ]
      }
    }
  },
  ...
}
```
**Case 3**: no permissions specified
```
userSync: {
  userIds: [{
    name: "sharedId",
    ...
    }
  ]
}
```
Rubicon bid adapter receives:
```
{
  "bidder": "rubicon",
  ...
  "userId": {
    "sharedid": {
      "id": "01EQDMBY1YPFFVY57R2V11PY89",
      "third": "01EQDMBY1YPFFVY57R2V11PY89"
    }
  },
  "userIdAsEids": [
    {
      "source": "sharedid.org",
      "uids": [
        {
          "id": "01EQDMBY1YPFFVY57R2V11PY89",
          "atype": 1,
          "ext": {
            "third": "01EQDMBY1YPFFVY57R2V11PY89"
          }
        }
      ]
    }
  ],
 ...
}
```
PrebidServerBidAdapter request:
```
{
  ...
  "ext": {
    "prebid": {
      ...
      "data": {
         ...
      }
    }
  },
  ...
  "user": {
    "ext": {
      "eids": [
        {
          "source": "sharedid.org",
          "uids": [
            {
              "id": "01EQDMBY1YPFFVY57R2V11PY89",
              "atype": 1,
              "ext": {
                "third": "01EQDMBY1YPFFVY57R2V11PY89"
              }
            }
          ]
        }
      ]
    }
  }
}
```
@jdwieland8282

### CODE REVIEW

**Comment by bretg** [comment]
This is going to need update to modules/prebidServerBidAdapter as well.

We need to send the eidPermission in openrtb. There are a couple of ways to do this, but maybe the easiest would be to have pbsBidAdapter call a function of the userId module to get the json block?

File: modules/prebidServerBidAdapter/index.js
Comment (bretg) [inline modules/prebidServerBidAdapter/index.js:640]: I may be misunderstanding this, but it's not allowed to pass the eidPermissions directly to bidders.

i.e if this implies that the eidPermissions block is passed around on the bidRequest object, that would let client-side adapters know where they stand.

Comment (YerkovichM) [reply modules/prebidServerBidAdapter/index.js:640]: Ok, I`ll change it

File: modules/userId/eids.js
Comment (mmoschovas) [inline modules/userId/eids.js:2]: utils.js is already imported above. Can you remove this second declaration and use the utils var instead?

File: modules/userId/index.js
Comment (mmoschovas) [inline modules/userId/index.js:127]: Similar to the other comment, eids.js is defined above. Please combine as : 

`import  {createEidsArray, getEidPermissions } from './eids.js'`

**Review by mmoschovas — CHANGES_REQUESTED** [review]
The above two changes should fix the circleci errors, but I think there should be a unit test added to prebidServerBidAdapter_spec to validate these changes.

**Comment by YerkovichM** [comment]
This should be a functional part. Unit tests are coming next

**Comment by YerkovichM** [comment]
@mmoschovas this should be it. Review the PR please

**Review by mmoschovas — APPROVED** [review]
Thanks for adding the tests. LGTM

File: modules/userId/eids.js
Comment (smenzer) [inline modules/userId/eids.js:230]: since this list of bidders is being sent to SSPs and likely will land in OpenRTB bid requests, should the name `bidders` in the eids object be changed to `pbjs_bidders` since it won't really apply to the openrtb world?

also, rather than `['*']` can we just omit `bidders` (or `pbjs_bidders`) altogether if none were set? This will avoid sending a field that isn't necessary

**Review by smenzer — CHANGES_REQUESTED** [review]
can you please add a test to cover the following scenario?
* `bidders` is set but doesn't include `sampleBidder`; check that the ID is not available in their bid request, but is available in the bidders listed

File: modules/userId/eids.js
Comment (YerkovichM) [reply modules/userId/eids.js:230]: @jdwieland8282

Comment (jdwieland8282) [reply modules/userId/eids.js:230]: @smenzer changing bidders to pbjs_bidders seems fine to me. And yeah I agree omitting `pbjs_bidders` if none are set makes sense. Thanks.

**Comment by jdwieland8282** [comment]
Docs update: https://github.com/prebid/prebid.github.io/pull/2592

**Comment by YerkovichM** [comment]
@smenzer I commited the update. Added test and renamed bidders to pbjs_bidders

**Review by smenzer — APPROVED** [review]
LGTM

**Comment by smenzer** [comment]
@mmoschovas can you please have another check of the new changes and then you can merge if you're happy.

**Comment by jdwieland8282** [comment]
@mmoschovas or @smenzer can this be merged in?

**Comment by bretg** [comment]
I'd like to suggest that `pbjs_bidders` become `pbjsBidders` or better yet just `bidders`.

There's no precedence in PBJS for attributes with an underscore. It's always camel-case.

**Comment by jdwieland8282** [comment]
@YerkovichM mind updating with @bretg 's comment above regarding the naming of `pbjs_bidders` to `bidders`

**Comment by smenzer** [comment]
Camel case is fine with me but "bidders" in openrtb has a meaning (ie DSP) and in this case we're referring to prebid bidder adapters, which are almost exclusively SSPs. So I suggest we stick with pbjsBidders to make it clear what the value will be for

**Comment by smenzer** [comment]
LGTM. since this touches core, @mmoschovas can you please do a final review with the updated code changes and then you can merge if you approve?

**Comment by bretg** [comment]
>  "bidders" in openrtb has a meaning (ie DSP) and in this case we're referring to prebid bidder adapters, which are almost exclusively SSPs

I'd like to disagree with the decision to call this "pbjsBidders". 

1) bidder codes are the same between pbjs and pbs.
2) we've already got this feature under code review in PBS-Java per the original spec:

{code}
{
    ext: {
        prebid: {
            data: {
                eidpermissions: [   // prebid server will use this to filter user.ext.eids
                   {"source": "sharedid.org", "bidders": ["*"]},
                   {"source": "neustar.biz", "bidders": ["bidderB"]},
                   {"source": "id5-sync.com", "bidders": ["bidderA","bidderC"]}
                ]
            }
        }
    }
}
{code}

I'd like to stick with the original plan please?

File: modules/prebidServerBidAdapter/index.js
Comment (bretg) [inline modules/prebidServerBidAdapter/index.js:662]: This needs to be lowercase eidpermissions.

File: modules/prebidServerBidAdapter/index.js
Comment (bretg) [inline modules/prebidServerBidAdapter/index.js:658]: please make this 'bidders' and not 'pbjsBidders'.

**Review by bretg — CHANGES_REQUESTED** [review]

**Comment by smenzer** [comment]
> I'd like to stick with the original plan please?

@bretg is this feature intended only for prebid server? i see your point w.r.t. naming for PBS only.

**Comment by YerkovichM** [comment]
@bretg updated

**Comment by smenzer** [comment]
@Fawke can I get your second review since this touches core, please?

**Review by Fawke — APPROVED** [review]
Looks good, have tested this with few adapters, it works as expected.

**Comment by bretg** [comment]
> is this feature intended only for prebid server?

No.

BTW - I updated the original description to reflect the current understanding: `bidders` and not `pbjs_bidders`.

Note that I also changed the expected results of Case 3 -- if `bidders` isn't specified, there's no need to send ext.prebid.data.eidpermissions to PBS at all. But looking at the code, I suspect that's not what's happening. Looks like ext.prebid.data.eidpermissions will be there every time whether there are special permissions or not. @YerkovichM , please confirm?  

I would like to propose we change this behavior... if a given EID doesn't specify `bidders` I don't think it should be part of the output of buildEidPermissions.

File: modules/userId/eids.js
Comment (bretg) [inline modules/userId/eids.js:237]: I don't think we should be adding anything to the eidPermissions array unless there are actually permissions to be added. i.e. if there's no `bidders` array specified, why do we need an entry at all?

Comment (YerkovichM) [reply modules/userId/eids.js:237]: Previously entry contained bidders: ['*'] if every bidder allowed, but then decided to omit bidders in this case.
Anyway, I will omit this entry entirely

**Comment by YerkovichM** [comment]
@bretg now if a given EID doesn't specify bidders it is not returnd by buildEidPermissions and therefore it is not incuded in prebid 
server request

**Review by bretg — APPROVED** [review]

**Comment by bretg** [comment]
@YerkovichM - looks ok to me now. Thanks. Fix conflicts please, and then we can merge.

**Comment by YerkovichM** [comment]
@bretg conflicts are resolved

**Pull request merged**
